tls/s2n_x509_validator.c

```diff
@@ -436,26 +436,59 @@ S2N_RESULT s2n_x509_validator_check_cert_preferences(struct s2n_connection *conn
      *# "signature_algorithms" extension also applies to signatures appearing in
      *# certificates.
      */
-    if (security_policy->certificate_signature_preferences == NULL) {
+    if (security_policy->certificate_signature_preferences == NULL
+            && security_policy->certificate_key_preferences == NULL) {
         return S2N_RESULT_OK;
     }
 
     struct s2n_cert_info info = { 0 };
     RESULT_GUARD(s2n_openssl_x509_get_cert_info(cert, &info));
 
-    if (info.self_signed) {
-        return S2N_RESULT_OK;
-    }
-
     /* Ensure that the certificate signature does not use SHA-1. While this check
      * would ideally apply to all connections, we only enforce it when certificate
      * preferences exist to stay backwards compatible.
      */
-    if (conn->actual_protocol_version == S2N_TLS13) {
+    if (!info.self_signed && conn->actual_protocol_version == S2N_TLS13) {
         RESULT_ENSURE(info.signature_digest_nid != NID_sha1, S2N_ERR_CERT_UNTRUSTED);
     }
 
-    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    if (!info.self_signed) {
+        RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    }
+    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_key(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+
+    return S2N_RESULT_OK;
+}
+
+/* Validates that the root certificate uses a key allowed by the security policy
+ * certificate preferences.
+ */
+static S2N_RESULT s2n_x509_validator_check_root_cert(struct s2n_x509_validator *validator, struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(conn);
+
+    const struct s2n_security_policy *sp = NULL;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &sp));
+
+    if (!sp->certificate_preferences_apply_locally) {
+        return S2N_RESULT_OK;
+    }
+
+    RESULT_ENSURE_REF(validator->store_ctx);
+    DEFER_CLEANUP(STACK_OF(X509) *cert_chain = X509_STORE_CTX_get1_chain(validator->store_ctx),
+            s2n_openssl_x509_stack_pop_free);
+    RESULT_ENSURE_REF(cert_chain);
+
+    const int certs_in_chain = sk_X509_num(cert_chain);
+    RESULT_ENSURE(certs_in_chain > 0, S2N_ERR_CERT_UNTRUSTED);
+    X509 *root = sk_X509_value(cert_chain, certs_in_chain - 1);
+    RESULT_ENSURE_REF(root);
+
+    struct s2n_cert_info info = { 0 };
+    RESULT_GUARD(s2n_openssl_x509_get_cert_info(root, &info));
+
+    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_key(sp, &info), S2N_ERR_CERT_UNTRUSTED);
```

Comment:
I think usage error makes more since. And I agree that `S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT` is more descriptive, so I went ahead and switched it to that.

How stingy are we with adding new Err's? The current description for that err is
```
"Incompatibility found between loaded certificates and chosen security policy"
```
which is technically accurate, but I think it might be nice to have a more explicit `S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_ROOT` error for this situation, since I would imagine that some customers might not be thinking of the certs in their trust store as "loaded".

Reply:
Hm I feel like it does makes sense to describe certs as being "loaded" into the trust store, but that could definitely just be me. I don't think it would be bad to add a new error if you think it would be more clear.

If you do add a new error I would maybe go with something like INCOMPATIBLE_TRUSTED_CERT/TRUST_STORE_CERT since a cert in the trust store doesn't have to be a root.